### PR TITLE
cqlsh: try server-side DESCRIBE, then client-side 

### PR DESCRIPTION
**Pull request by sylwiaszunejko**
Since Scylla 5.2 (https://github.com/scylladb/scylladb/commit/e6ffc220539a05cf7b2a0e37368c27075d809476) a support for server-side DESCRIBE was added. However, cqlsh did not start to use this functionality, since it is only enabled if it detects CQL version at least 4. Scylla did not increase this version number as it doesn't support all of its features, so there is a need for a different detection mechanism for server-side DESCRIBE.

This PR changes the behavior in do_describe: cqlsh will first try to execute the server-side DESCRIBE. If this fails with SyntaxException, meaning that Scylla doesn't support that command, cqlsh falls back to the client-side DESCRIBE behavior. Also I fixed tests to be in line with recent changes to output sent by Scylla.

The other possible solutions were rejected:
- Based on Scylla version: would require ugly hard-coding of versions
- Modifying Scylla to provide some indication that this feature is enabled: Scylla 5.2 is already released without it, by implementing it in another way, we'll get it out sooner
- Do a trial "DESCRIBE" at the start of connection to detect if the server supports it: if cqlsh ever supported connecting to multiple nodes (right now it uses WhiteListRoundRobinPolicy) we would have to do the check on all of the nodes in case a rolling upgrade is currently occurring and some of the nodes don't support server-side DESCRIBE.

The change was tested manually on a couple of last Scylla OSS.

Fixes https://github.com/scylladb/scylla-cqlsh/issues/17

### CODE REVIEW

**Review by avelanarius — APPROVED** [review]

**Comment by avelanarius** [comment]
One nit: "[Fix tests](https://github.com/scylladb/scylla-cqlsh/pull/88/commits/3c28b6a2185a177c06a06e5d54b16d02bf4c104a)" commit should be a part of the last commit - either @sylwiaszunejko should squash it or a scylla-cqlsh maintainer should merge it with squashing.

File: bin/cqlsh.py
Comment (piodul) [inline bin/cqlsh.py:1500]: The name of the `client_side` parameter implies that it controls whether client-side describe will be used or not. It's not entirely true because in case of unsupported `client_side=False` we will fall back to client-side describe. How about `force_client_side_describe`?

Moreover, I'm not sure whether this should be a public parameter. The upstream doesn't allow to control whether client- or server-side describe is used. Perhaps we should turn this into a private function and keep a public `do_describe` without the additional parameter.

**Comment by piodul** [comment]
@fruch There are some dtests that use cqlsh to run DESCRIBE. Do we need to run them to see which dtests need fixing before this is merged?

**Comment by piodul** [comment]
cc: @Jadw1 - please review as well

**Comment by fruch** [comment]
> @fruch There are some dtests that use cqlsh to run DESCRIBE. Do we need to run them to see which dtests need fixing before this is merged?

Yes we, otherwise we won't pass gating and won't be able to pass this change into scylla core

And we should be aware of those small differences in output, some of them might need fixes to the core, and might break tools/users

**Comment by Jadw1** [comment]
> > @fruch There are some dtests that use cqlsh to run DESCRIBE. Do we need to run them to see which dtests need fixing before this is merged?
> 
> Yes we, otherwise we won't pass gating and won't be able to pass this change into scylla core
> 
> And we should be aware of those small differences in output, some of them might need fixes to the core, and might break tools/users

Are those dtests using cqlsh or 'cql.execute()'? Second option is fine since the statement is send directly to Scylla.
I'm not aware of all dtests, but I'm just pointing out the difference.

**Comment by fruch** [comment]
> > > @fruch There are some dtests that use cqlsh to run DESCRIBE. Do we need to run them to see which dtests need fixing before this is merged?
> > 
> > Yes we, otherwise we won't pass gating and won't be able to pass this change into scylla core
> > 
> > And we should be aware of those small differences in output, some of them might need fixes to the core, and might break tools/users
> 
> Are those dtests using cqlsh or `cql.execute()`? Second option is fine since the statement is send directly to Scylla.
> I'm not aware of all dtests, but I'm just pointing out the difference.

They are using cqlsh, no one wrote dtest for server side describe

**Comment by sylwiaszunejko** [comment]
@fruch How can I change `cqlsh` version in dtest to test my changes? And are these the tests you wrote about https://github.com/scylladb/scylla-dtest/blob/next/cqlsh_tests/cqlsh_tests.py?

File: bin/cqlsh.py
Comment (sylwiaszunejko) [reply bin/cqlsh.py:1500]: I am not sure how this private function should work. Would it work to add a new parameter (force_client_side_describe = False) to the Shell class, set it to True before executing do_describe if the server doesn't support the DESCRIBE query, and then revert it back to False to attempt server-side by default?

**Comment by fruch** [comment]
> @fruch How can I change `cqlsh` version in dtest to test my changes? And are these the tests you wrote about https://github.com/scylladb/scylla-dtest/blob/next/cqlsh_tests/cqlsh_tests.py?

You need to change the submodule in scylla core, and build the scylla unified package, then you can use that in dtest.

File: pylib/cqlshlib/test/test_cqlsh_output.py
Comment (fruch) [reply pylib/cqlshlib/test/test_cqlsh_output.py:687]: we should be able to fallback as needed, user can be using cqlsh directy by it's docker (i.e. scylla-cloud point to it)
or if they retrieved it by `pip install` we can't assume server-side is there

File: bin/cqlsh.py
Comment (fruch) [reply bin/cqlsh.py:1500]: regardless of compatibility with upstream, which is already broken, since we introduce the client side back (I did it when we start working on this repo)

I think it would be nice to have an escape hatch for users, not sure how many would really need it.

Comment (piodul) [reply bin/cqlsh.py:1500]: I was thinking about:

```python
def do_describe(self, parsed):
    self._do_describe(force_client_side_describe=False)

def _do_describe(self, force_client_side_describe):
    # ...
        except cassandra.protocol.SyntaxException:
            self._do_describe(force_client_side_describe=True)
```

Comment (piodul) [reply bin/cqlsh.py:1500]: > I think it would be nice to have an escape hatch for users, not sure how many would really need it.

In that case, the transformation I suggested shouldn't be applied.

I still think that the parameter name `client_side` should be changed to `force_client_side_describe`, though.

Comment (fruch) [reply bin/cqlsh.py:1500]: the suggestion would drift us more than the upstream, so yes it's not really that critical.
yes, the suggestion is a bit clearer, even that that not force involved here :), but any "bool" like prefix would work (use/is), and the suffix of describe, well.. that the name of the function, so might not be totally needed

**Review by Jadw1 — APPROVED** [review]

File: pylib/cqlshlib/test/test_cqlsh_output.py
Comment (nyh) [reply pylib/cqlshlib/test/test_cqlsh_output.py:687]: I agree with @piodul:
1. This commit will be on the git log with the mysterious commit message "fix tests" :-) What was fixed and why?
2. How can the test assume that it is talking to a new Scylla? I think it needs to allow both old and new formats.

File: pylib/cqlshlib/test/test_cqlsh_output.py
Comment (nyh) [inline pylib/cqlshlib/test/test_cqlsh_output.py:705]: Seeing these changes made me convinced that the way this test is written should be changed. It shouldn't compare the full string, but rather the content. I mean, what if in the future we decide to sort the column names differently (I think there's a discussion about that right now? To format the WITH below differently, and so on? It will break this test.

By the way, won't this test break every time we add a new WITH parameter to Scylla? I already see you added "cdc" and "paxos_grace_seconds", we'll need to add more and more of those :-( Are we fine with that?

Oh, and don't we allow someone to use cqlsh with Cassandra? It seems this test won't work with Cassandra.

I don't know what to recommend instead of this method. I can think of two ideas:
1. Use "describe" to get the description of a table and then create this table again and "describe" it again and check the two descriptions are identical.
2. Compare the expected strings after passing them through some sort of "canonization" function, which removes whitespace, sorts columns and WITH options, canonizes floating point values, etc. This won't be easy. I don't know what to recommend about non-standard options like cdc and paxos_grace_seconds.

File: bin/cqlsh.py
Comment (sylwiaszunejko) [reply bin/cqlsh.py]: @piodul @fruch I'm not sure which version I should choose, the initial one or the one proposed by @piodul

File: pylib/cqlshlib/test/test_cqlsh_output.py
Comment (sylwiaszunejko) [reply pylib/cqlshlib/test/test_cqlsh_output.py:705]: > By the way, won't this test break every time we add a new WITH parameter to Scylla? 

Probably, but I am not sure if refactoring these tests should be a part of this PR, maybe we should create a separate issue to change it?

> Oh, and don't we allow someone to use cqlsh with Cassandra? It seems this test won't work with Cassandra.

In `test_describe_columnfamily_output` there are two different versions of expected output, we use different one if scylla is used and I only changed that part so it should work fine with Cassandra.

File: pylib/cqlshlib/test/test_cqlsh_output.py
Comment (sylwiaszunejko) [reply pylib/cqlshlib/test/test_cqlsh_output.py:687]: I updated the commit message

**Comment by sylwiaszunejko** [comment]
I have run the dtests and there are 9 tests failing with my changes: 
```
FAILED cqlsh_tests/cqlsh_tests.py::TestCqlshWithSSL::test_int_values - AssertionError: Output 
FAILED cqlsh_tests/cqlsh_tests.py::TestCqlshWithSSL::test_describe - assert ['CREATE KEYS...d, col)', ...] == ['CRE...
FAILED cqlsh_tests/cqlsh_tests.py::TestCqlshWithSSL::test_describe_mv - AssertionError: assert ['CREATE MATE...me ASC)...
FAILED cqlsh_tests/cqlsh_tests.py::TestCqlsh::test_int_values - AssertionError: Output 
FAILED cqlsh_tests/cqlsh_tests.py::TestCqlsh::test_describe - assert ['CREATE KEYS...d, col)', ...] == ['CRE...
FAILED cqlsh_tests/cqlsh_tests.py::TestCqlsh::test_describe_mv - AssertionError: assert ['CREATE MATE...me ASC)...
FAILED cqlsh_tests/cqlsh_tests.py::TestCqlLogin::test_login_keeps_keyspace - AssertionError: assert ['ks1table'] == ['ks1ta...
```
I will now try to fix it.

**Comment by fruch** [comment]
> I have run the dtests and there are 9 tests failing with my changes:
> 
> ```
> FAILED cqlsh_tests/cqlsh_tests.py::TestCqlshWithSSL::test_int_values - AssertionError: Output 
> FAILED cqlsh_tests/cqlsh_tests.py::TestCqlshWithSSL::test_describe - assert ['CREATE KEYS...d, col)', ...] == ['CRE...
> FAILED cqlsh_tests/cqlsh_tests.py::TestCqlshWithSSL::test_describe_mv - AssertionError: assert ['CREATE MATE...me ASC)...
> FAILED cqlsh_tests/cqlsh_tests.py::TestCqlsh::test_int_values - AssertionError: Output 
> FAILED cqlsh_tests/cqlsh_tests.py::TestCqlsh::test_describe - assert ['CREATE KEYS...d, col)', ...] == ['CRE...
> FAILED cqlsh_tests/cqlsh_tests.py::TestCqlsh::test_describe_mv - AssertionError: assert ['CREATE MATE...me ASC)...
> FAILED cqlsh_tests/cqlsh_tests.py::TestCqlLogin::test_login_keeps_keyspace - AssertionError: assert ['ks1table'] == ['ks1ta...
> ```
> 
> I will now try to fix it.

if you are fixing the dtest, I would suggest doing it in a way it would work for both (so we can merge before, and not force us to merge it at the same time)

**Comment by sylwiaszunejko** [comment]
The PR with changes to the dtests is now merged https://github.com/scylladb/scylla-dtest/pull/4319 and I did some refactoring of the tests here to make it work for different formatting versions.
@fruch @avelanarius @piodul

File: bin/cqlsh.py
Comment (fruch) [reply bin/cqlsh.py]: I think the current name is good enough

File: pylib/cqlshlib/test/test_cqlsh_output.py
Comment (fruch) [reply pylib/cqlshlib/test/test_cqlsh_output.py]: now the test are a bit more easy to grok, and it's o.k. from my POV

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**
